pkg/bundle/publish/publish.go

```diff
@@ -4,38 +4,151 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strings"
 
+	"github.com/RedHatGov/bundle/pkg/archive"
+	"github.com/RedHatGov/bundle/pkg/config"
+	"github.com/RedHatGov/bundle/pkg/config/v1alpha1"
+	"github.com/RedHatGov/bundle/pkg/image"
+	"github.com/RedHatGov/bundle/pkg/metadata/storage"
+	"github.com/google/uuid"
+	"github.com/mholt/archiver/v3"
 	"github.com/openshift/oc/pkg/cli/image/mirror"
 	"github.com/sirupsen/logrus"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
-
-	"github.com/RedHatGov/bundle/pkg/archive"
-	"github.com/RedHatGov/bundle/pkg/config"
-	"github.com/RedHatGov/bundle/pkg/image"
 )
 
-// TODO: refactor into Complete() -> Validate() -> Run() CLI pattern (thing `oc`).
+type UuidError struct {
+	InUuid   uuid.UUID
+	CurrUuid uuid.UUID
+}
+
+func (u *UuidError) Error() string {
+	return fmt.Sprintf("Mismatched UUIDs. Want %v, got %v", u.CurrUuid, u.InUuid)
+}
+
+type SequenceError struct {
+	inSeq   int
+	CurrSeq int
+}
+
+func (s *SequenceError) Error() string {
+	return fmt.Sprintf("Bundle Sequence out of order. Current sequence %v, incoming sequence %v", s.CurrSeq, s.inSeq)
+}
 
 func (o *Options) Run(ctx context.Context) error {
 
+	var currentMeta v1alpha1.Metadata
+	var incomingMeta v1alpha1.Metadata
+	a := archive.NewArchiver()
+
 	logrus.Infof("Publish image set from archive %q to registry %q", o.ArchivePath, o.ToMirror)
 
-	// Load tar archive made by `create`.
-	if err := archive.NewArchiver().Unarchive(o.ArchivePath, o.Dir); err != nil {
+	// Create workspace
+	tmpdir, err := ioutil.TempDir(o.Dir, "imageset")
+	if err != nil {
+		return err
+	}
+
+	logrus.Debugf("Created temporary directory %s", tmpdir)
+	defer os.RemoveAll(tmpdir)
+
+	// Get file information from the source archives
+	filesInArchive, err := o.readImageSet(a)
+	if err != nil {
+		return err
+	}
+
+	target := filepath.Join(config.PublishDir, config.MetadataFile)
+	dest := filepath.Join(o.Dir, target)
+
+	// Create backend for o.Dir
+	backend, err := storage.NewLocalBackend(o.Dir)
+	if err != nil {
+		return fmt.Errorf("error opening local backend: %v", err)
+	}
+
+	// Create a local workspace backend
+	workspace, err := storage.NewLocalBackend(tmpdir)
+	if err != nil {
+		return fmt.Errorf("error opening local backend: %v", err)
+	}
+
+	// Check for existing metadata. Metadata will be extracted before
```

Comment:
Can you walk me through why local metadata would exist, and even if it did, the CLI should use it? A most-recent metadata file is already packed in the archive and is consistent with what `publish` expects (unless it was modified in transit somehow).

Reply:
@estroz Sure! So the publisher should have knowledge of all publishing runs so it will know if the bundle being passed is in the right sequence and has the correct UID. Right now the metadata is stored locally, but eventually I would like to have a general attempt to retrieve it from a backend on the disconnected side.